pkg/parca/parca.go

```diff
@@ -584,35 +584,43 @@ func runScraper(
 		},
 	)
 
-	parcaserver := server.NewServer(reg, version)
-	gr.Add(
-		func() error {
-			return parcaserver.ListenAndServe(
-				ctx,
-				logger,
-				flags.Port,
-				flags.CORSAllowedOrigins,
-				flags.PathPrefix,
-				server.RegisterableFunc(func(ctx context.Context, srv *grpc.Server, mux *runtime.ServeMux, endpoint string, opts []grpc.DialOption) error {
-					scrapepb.RegisterScrapeServiceServer(srv, m)
-					if err := scrapepb.RegisterScrapeServiceHandlerFromEndpoint(ctx, mux, endpoint, opts); err != nil {
-						return err
-					}
-					return nil
-				}),
-			)
-		},
-		func(_ error) {
-			ctx, cancel := context.WithTimeout(ctx, 30*time.Second) // TODO make this a graceful shutdown config setting
-			defer cancel()
+	{
+		parcaserver := server.NewServer(reg, version)
+		ctx, cancel := context.WithCancel(ctx)
+		gr.Add(
+			func() error {
+				return parcaserver.ListenAndServe(
+					ctx,
+					logger,
+					flags.Port,
+					flags.CORSAllowedOrigins,
+					flags.PathPrefix,
+					server.RegisterableFunc(func(ctx context.Context, srv *grpc.Server, mux *runtime.ServeMux, endpoint string, opts []grpc.DialOption) error {
+						scrapepb.RegisterScrapeServiceServer(srv, m)
+						if err := scrapepb.RegisterScrapeServiceHandlerFromEndpoint(ctx, mux, endpoint, opts); err != nil {
+							return err
+						}
+						return nil
+					}),
+				)
+			},
+			func(_ error) {
+				level.Debug(logger).Log("msg", "server shutting down")
 
-			level.Debug(logger).Log("msg", "server shutting down")
-			err := parcaserver.Shutdown(ctx)
-			if err != nil && !errors.Is(err, context.Canceled) {
-				level.Error(logger).Log("msg", "error shutting down server", "err", err)
-			}
-		},
-	)
+				// Cancels ListenAndServe.
+				cancel()
```

Comment:
we should only call this cancel after we've exceeded the graceful timeout, otherwise, requests may error in context canceled even if they would have completed within the graceful shutdown window

Reply:
Done.